app/src/main/java/com/github/se/cyrcle/ui/map/MapScreen.kt

```diff
@@ -242,101 +241,116 @@ fun MapScreen(
+                          options =
+                              viewAnnotationOptions {
+                                annotatedLayerFeature(LAYER_ID_RECT) {
+                                  featureId(it.id)
+                                  geometry(it.geometry)
+                                  annotationAnchor {
+                                    anchor(ViewAnnotationAnchor.BOTTOM)
+                                    offsetY(0.0)
+                                  }
+                                }
+                              })
+
+                  // Set the text and the button of the view annotation
+                  ItemCalloutViewBinding.bind(viewAnnotation).apply {
+                    textNativeView.text =
+                        screenCapacityString.format(parkingDeserialized.capacity.description)
+                    selectButton.setOnClickListener {
+                      parkingViewModel.selectParking(parkingDeserialized)
+                      navigationActions.navigateTo(Screen.PARKING_DETAILS)
+                    }
                   }
+                  true
+                })
+            // ======================= RECTANGLES CLICK =======================
 
-                  // store the zoom level
-                  // This must stay at the end of the listener.
-                  zoomState.value = mapView.mapboxMap.cameraState.zoom
-                }
+            // =======================  CAMERA  LISTENER  =======================
+            mapView.mapboxMap.subscribeCameraChanged {
+
+              // Get the top right and bottom left coordinates of the current view only when
+              // what the user sees is outside the screen
+              val (currentBottomLeft, currentTopRight) = mapViewModel.getScreenCorners(mapView)
+
+              // Temporary fix to avoid loading too much parkings when zoomed out
+              if (mapView.mapboxMap.cameraState.zoom > thresholdDisplayZoom) {
+                parkingViewModel.getParkingsInRect(currentBottomLeft, currentTopRight)
+              }
+              // On zoom-out past the threshold, switch to the markers mode
+              if (mapView.mapboxMap.cameraState.zoom < ADVANCED_MODE_ZOOM_THRESHOLD &&
+                  zoomState.value >= ADVANCED_MODE_ZOOM_THRESHOLD) {
+                mapMode.value = MapMode.MARKERS
+              }
+              // On zoomin in past the threshold, switch to the user's selected mode
+              if (mapView.mapboxMap.cameraState.zoom >= ADVANCED_MODE_ZOOM_THRESHOLD &&
+                  zoomState.value < ADVANCED_MODE_ZOOM_THRESHOLD) {
+                mapMode.value = userMapMode.value
+              }
+
+              // store the zoom level
+              // This must stay at the end of the listener.
+              zoomState.value = mapView.mapboxMap.cameraState.zoom
+            }
             // =======================  CAMERA  LISTENER  =======================
 
             onDispose {
```

Comment:
Shouldn't we call `removePreviewCard` and set the annotation manager to null in the ViewModel on dispose?

Reply:
Yes we could, that would not save a lot of memory but let me add it anyway